examples/adrv9002-iiostream.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * libiio - Library for interfacing industrial I/O (IIO) devices
+ *
+ * Copyright (C) 2021 Analog Devices, Inc.
+ * Author: Nuno Sá <nuno.sa@analog.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ * */
+#define _BSD_SOURCE
+#define _DEFAULT_SOURCE
+#include <iio.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#define error(fmt, ...) \
+	printf("%s, %d: "fmt, __func__, __LINE__, ##__VA_ARGS__)
+
+#define info(fmt, ...) \
+	printf("%s, %d: "fmt, __func__, __LINE__, ##__VA_ARGS__)
```

Comment:
OSX doesn't like your `##__VA_ARGS__`...

Reply:
Arghhh, I guess I will also have to drop my neat cleanup handling?

https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=4022&view=logs&j=afc9f219-4f50-5293-48d5-d83ea574007e&t=470c08d8-8808-547a-c046-cb3882e7f5bd&l=185

Reply:
Shouldn't you use `atexit()`?

Reply:
Hmm the reason why I did not used it is because I cannot pass any argument to the cb... But for this example I can always make the objects global and that solves it! I will update it together with the variadic macro...

Reply:
Well, it actually makes it very cumbersome with `atexit()`. I will just drop this and add a normal cleanup function...

Reply:
Have a look at iio-debug.h in the dev branch, there are macros that basically do the same thing and are cross-platform.

Reply:
Just gave it a look but I do not think I can use it directly without some tweaks. The thing is that on the format I have, I want to prepend `__func__` and `__LINE__` to my format string. So, I cannot guarantee that `__VA_ARGS__` will expand to something as you can [here](https://github.com/analogdevicesinc/libiio/blob/dev/iio-debug.h#L85) with `buf`. First thing that comes to my mind is to just add a dummy empty string at the end of my format so that I can nicely extract my arguments. The other one is just using two printfs like:

```
#define error(...) \
        printf("%s, %d: ERROR: ", __func__, __LINE__); \
        printf(__VA_ARGS__);
```
Anyways, I just don't want to wast to much time on this... As I don't have HW with me today I will check this tomorrow. I want to do a sanity check before pushing this update...

Reply:
```
#define __error(line, ...) printf(__func__ ", " #line __VA_ARGS__);
#define _error(line, ...) __error(line, __VA_ARGS__);
#define error(fmt, ...) _error(__LINE__, fmt, __VA_ARGS__);
```

That doesn't work?

Reply:
Nope, I made a quick test and it does not compile. I'm not sure we can have `printf(__func__"plus something")`... What I have in mind for the dummy string is something like:

```
#define ARGS(fmt, ...)  __VA_ARGS__
#define FMT(fmt, ...)   fmt
#define error(...) \
       printf("%s, %d: ERROR: " FMT(__VA_ARGS__)"%s", __func__, __LINE__, ARGS(__VA_ARGS__, ""))
```

I could also just use `\n` instead of the empty string. I do not like much to enforce that but as this is an example app I don't really mind (if you prefer it...)

Reply:
@nunojsa 
```
#define __FIRST(a, ...)		a
#define ___OTHERS(a, ...)	a, __VA_ARGS__
#define __OTHERS(a, b, ...)	___OTHERS(a, __VA_ARGS__)
#define error(...) \
	printf("%s, %d: ERROR: " __FIRST(__VA_ARGS__, 0), __func__, __OTHERS(__LINE__,__VA_ARGS__))
```

Reply:
Does not works for all cases:

```
int n = 10;
error("Hello world: %d\n", n); // works fine
error("Hello world\n"); //compilation error
```
I think the issue with the later is that __VA_ARGS__ will evaluate to nothing in `___OTHERS` which is not nicely handled unless we use the gnu extension. That is why I think the "dummy" string after the format is needed (at least is one way to handle it; maybe not the best one).

Out of curiosity, is there any special reason for you to use `__FIRST(__VA_ARGS__, 0)` (also saw it in iio-debug.h)? If I'm not mistaken, `__FIRST(__VA_ARGS__)` should also work...

Reply:
C requires that the number of parameters in a macro's "..." is >= 1. iio-debug.h does `__FIRST(__VA_ARGS__, 0)` otherwise the compiler complains.

Reply:
Following your suggestion:
```
#include <stdio.h>

#define __FIRST(a, ...)		a
#define ___OTHERS(a, ...)	a, __VA_ARGS__
#define __OTHERS(a, b, ...)	___OTHERS(a, __VA_ARGS__)

#define error(...) \
	printf("%s, %d: ERROR: " __FIRST(__VA_ARGS__, 0) "%c", __func__, __OTHERS(__LINE__,__VA_ARGS__, '\0'))

int main(void)
{
	error("Test: %d\n", 5);
	error("Test2\n");
}
```

This seems to compile fine.

Reply:
> C requires that the number of parameters in a macro's "..." is >= 1. iio-debug.h does __FIRST(__VA_ARGS__, 0) otherwise the compiler complains.

Hmm, I see. I guess with gnu extensions this is not really the case. That is why gcc did not complained when I tried it...